platform/android/java/lib/src/org/godotengine/godot/GodotView.java

```diff
@@ -139,6 +140,7 @@ private void init(XRMode xrMode, boolean translucent, int depth, int stencil) {
 				 * is interpreted as any 32-bit surface with alpha by SurfaceFlinger.
 				 */
 				if (translucent) {
+					this.setZOrderOnTop(true);
```

Comment:
@pouleyKetchoupp Can you clarify what this does? According to the documentation:

> By setting this, you cause it to be placed above the window. This means that none of the contents of the window this SurfaceView is in will be visible on top of its surface.

Which would mean that it would prevent additional component from rendering on top of the surface; is that the desired behavior?

Could you also walk me through how per-pixel transparency is achieved; it may help me understand better the logic in the PR.
Thanks!

Reply:
> Which would mean that it would prevent additional component from rendering on top of the surface; is that the desired behavior?

Yes, this allows Godot rendering to be displayed on top, so a plugin can render behind the transparent surface, rather than always in front (like it does by default). As a side effect, plugins can't render on top of the surface when this is enabled, as you mentioned, but it's the only way I've found to achieve transparent rendering.

> Could you also walk me through how per-pixel transparency is achieved; it may help me understand better the logic in the PR.

Transparent rendering is achieved this way:
1. `per_pixel_transparency/allowed` & `per_pixel_transparency/enabled` are both set to true in Project Settings
2. `get_tree().get_root().transparent_bg = true;` is called at runtime to make the viewport transparent
3. At Android export time, `--translucent` is added to the command line (this step is needed on 3.x because `Main::setup()` is called after the surface is initialized, it won't be needed on master).
4. At Android runtime, `--translucent` command line is checked to create a surface with transparent capabilities (on master it will just check the project settings directly)
5. When the plugin is initialized, its own rendering frame is placed on top of the main frame by default, so this can be used to move it behind:
```
ViewGroup parentView = (ViewGroup)pluginFrame.getParent();
parentView.removeView(pluginFrame);
parentView.addView(pluginFrame, 0);
```
(from what I've seen, both this and `setZOrderOnTop(true)` are needed for transparency to work correctly)

The result is the plugin does its custom rendering (for example camera preview) and Godot draws on top of it with transparent background (for example to add UI elements).

Let me know if there's more I can clarify!

Reply:
Remember that transparent viewports aren't supported yet with the Vulkan renderer: https://github.com/godotengine/godot/issues/40651

Reply:
@pouleyKetchoupp Sorry for the delay, I was out for the past week.

Regarding ordering of the surface and of the plugins' views, we control that logic [here](https://github.com/godotengine/godot/blob/e44379612d09a7b134bdadb901de409db170268c/platform/android/java/lib/src/org/godotengine/godot/Godot.java#L390) and [here](https://github.com/godotengine/godot/blob/e44379612d09a7b134bdadb901de409db170268c/platform/android/java/lib/src/org/godotengine/godot/Godot.java#L400), so couldn't just have a flag that tells us if `translucent` is enabled, and changed the order of the views as shown in your sample code?
With that approach, would we still need to set the `setZOrderOnTop(...)` flag?

Also, when you say the surface is transparent, does that mean that any view(s) `behind` it will be visible? I'm trying to clarify here since making the background transparent in Godot may not mean the same as making the surface view background transparent (which would allow to see underlying views).

Reply:
No worries!

> Regarding ordering of the surface and of the plugins' views, we control that logic [here](https://github.com/godotengine/godot/blob/e44379612d09a7b134bdadb901de409db170268c/platform/android/java/lib/src/org/godotengine/godot/Godot.java#L390) and [here](https://github.com/godotengine/godot/blob/e44379612d09a7b134bdadb901de409db170268c/platform/android/java/lib/src/org/godotengine/godot/Godot.java#L400), so couldn't just have a flag that tells us if `translucent` is enabled, and changed the order of the views as shown in your sample code?
> With that approach, would we still need to set the `setZOrderOnTop(...)` flag?

I'm not sure why (it could be the way GL surfaces work), but both are needed for the system to work correctly.
I've tested again since last time and here's how it goes:
-`setZOrderOnTop` is needed for transparent rendering, without it the Godot surface renders with black background and any view behind it is not visible.
-Changing the order of views is not strictly needed if `setZOrderOnTop` is set, but it helps with inputs as it allows Godot inputs to be taken over views placed behind.

> Also, when you say the surface is transparent, does that mean that any view(s) `behind` it will be visible? I'm trying to clarify here since making the background transparent in Godot may not mean the same as making the surface view background transparent (which would allow to see underlying views).

Yes, it allows other views to be visible behind it.

Using this feature, you can also make a transparent app and see the device desktop in the background. But that requires to also change the style of the main Godot app to this:
```
<style name="GodotAppMainTheme" parent="@android:style/Theme.Black.NoTitleBar.Fullscreen">
     <item name="android:windowIsTranslucent">true</item>
     <item name="android:windowBackground">@android:color/transparent</item>
</style>
```

So far I haven't found a nice way to enable it in the export process or by code, but it would be a nice addition to this PR.

Reply:
> -`setZOrderOnTop` is needed for transparent rendering, without it the Godot surface renders with black background and any view behind it is not visible.

@pouleyKetchoupp Can you double-check that [setZOrderMediaOverlay](https://developer.android.com/reference/android/view/SurfaceView#setZOrderMediaOverlay(boolean)) doesn't have a similar behavior to `setZOrderOnTop` (i.e: can it be used instead?).

> -Changing the order of views is not strictly needed if `setZOrderOnTop` is set, but it helps with inputs as it allows Godot inputs to be taken over views placed behind.

Okay, can we do that automatically then when the `translucent` flag is set (like I described above), this way the plugin itself won't need to move its own view.

> Using this feature, you can also make a transparent app and see the device desktop in the background. But that requires to also change the style of the main Godot app to this

That could be possible if it's detectable early enough that the `translucent` flag was set. If it's, you can swap the theme that's being used [here](https://github.com/godotengine/godot/blob/e44379612d09a7b134bdadb901de409db170268c/platform/android/java/app/src/com/godot/game/GodotApp.java#L44), using the current theme for non-translucency, and using the (new) translucent theme when the flag is set.

Reply:
> Can you double-check that setZOrderMediaOverlay doesn't have a similar behavior to setZOrderOnTop (i.e: can it be used instead?).

No, I've just tested and with `setZOrderMediaOverlay` the Godot view still renders with black background and other views are hidden behind so it looks like `setZOrderOnTop` is needed.

> Okay, can we do that automatically then when the translucent flag is set (like I described above), this way the plugin itself won't need to move its own view.

Yeah it would be great to have a helper for that! Should it be optional though? Maybe something like a plugin function `is_view_behind()` or `get_view_order()`.

The reason is there are two possible states at the moment:
-Either you change the view order and you detect inputs in the godot view, but the plugin rendered behind doesn't get any input (can be useful for adding buttons on top of the plugin view)
-Or you don't change the view order and the plugin receives inputs, but not the godot view (can be useful if you want to render on top of a plugin but keep the plugin responsible for inputs)

Ideally input events would be properly passed from godot to other views behind when not consumed by godot, but that would require some extra work to manage inputs differently.

> That could be possible if it's detectable early enough that the translucent flag was set. If it's, you can swap the theme that's being used here, using the current theme for non-translucency, and using the (new) translucent theme when the flag is set.

Ok, I'll check if I can figure out how to do this. I assumed the app `onCreate` method would be too early to get the information, but maybe I can access the same `--translucent` command line argument that's used to initialize the view.

Reply:
> We could add a `shouldBeOnTop()` method to the `GodotPlugin` interface which defaults to `true` (since that's the current default behavior) in order to figure out where to place the view.
> The issue with the `get_view_order()` approach is that each plugin would need to be aware of all other plugins.
> If that sounds like a good approach, it should be done in a separate PR since the update would also apply to the `master` branch.

That sounds good! I'll create a new PR for this.

Reply:
@pouleyKetchoupp Do you have an example for the logic you use to update the theme?

In general, the theme needs to be set before even calling `setContentView(...)`.
So the logic should look something like:

```
public void onCreate(Bundle savedInstanceState) {
		if (hasTranslucentFlag())
                    setTheme(R.style.GodotAppTranslucentTheme);
                else
                    setTheme(R.style.GodotAppMainTheme);
		super.onCreate(savedInstanceState);
}
```

where `GodotAppTranslucentTheme` would be as you described above:
```
<style name="GodotAppTranslucentTheme" parent="@android:style/Theme.Black.NoTitleBar.Fullscreen">
     <item name="android:windowIsTranslucent">true</item>
     <item name="android:windowBackground">@android:color/transparent</item>
</style>
```

Reply:
Based on your example, even just setting the transparent theme without checking any flag like this:
```
public void onCreate(Bundle savedInstanceState) {
                setTheme(R.style.GodotAppTranslucentTheme);
		super.onCreate(savedInstanceState);
}
```
doesn't seem to work, unless windowIsTranslucent is set to true in the default theme as well.

It looks like this flag can't be changed once the app is being created, and it works by adding it in the default theme only because the splash theme inherits from the default theme (unless I'm missing something).

Reply:
> It looks like this flag can't be changed once the app is being created, and it works by adding it in the default theme only because the splash theme inherits from the default theme (unless I'm missing something).

@pouleyKetchoupp I thought you were still looking into this. If not, I'm good with approving and unblocking this PR, and attempting to update the app theme for translucency in another PR. For that approach, can you create an issue that documents what you already tried; I can then take a look and see if anything is missing.

Reply:
Sorry for the confusion! Yes let me finalize this PR and open an issue to solve the app theme thing separately.